e2e tests for temporal data - RISDEV-3848

hamo225 opened the pull request.

hamo225 committed:
Add e2e tests for temporal data page

RISDEV-3848

hamo225 committed:
Update e2e tests

Adds test for editing
Joins tests for adding, editing, deleting together (for now)

RISDEV-3848

hamo225 committed:
Update e2e tests

RISDEV-3848

hamo225 committed:
Cleanup e2e: remove before each

RISDEV-3848

VictorDelCampo commented:
Hey @hamo225, this is my review:

1. I think we need a new e2e test for the side navigation where we test it separate? Or we just keep also the navigation within this test but then I would expect to be two describe-layers:
    - test navigation to time boundaries using navigation starting on `/amending-laws/eli/bund/bgbl-1/2017/s419/2017-03-15/1/deu/regelungstext-1` and just check if you arrived at .`../temporal-data`
    - test the actual page including:
       - heading
       - Artikel 3Inkraftrete
       - the managing of the data itself meaning add/update/delete
2. For testing the heading I would use if possible something more readable like:
```
   await expect(
      page.getByRole("heading", { level: 1, name: "Artikel 1" }),
    ).toBeVisible()
```
3. Why do you need to do 
```
    await page.waitForResponse(
      (response) =>
        response.url().includes("timeBoundaries") && response.status() === 200,
    )
```
   Is the await saveButton.click() not enough for waiting? But this could be related to the next point

4. Why do you do the API calls to test the response JSON? In my opinion the E2E tests are done user-facing meaning what the user would se. I would just add/update/delete data, save, reload page and see if the data is there

hamo225 committed:
Adjusts e2e

RISDEV-3848

reckseba committed:
Merge branch 'main' into risdev-3848

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+
+    await expect(dateInputs).toHaveCount(1)
+    const inputValue = await dateInputs.inputValue()
     await expect(inputValue).toBe("16.03.2017")
```
No `await` needed, because `toBe` does not return a promise.

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+    await page.reload()
+
+    const editedValue = await dateInputs.nth(1).inputValue()
+    await expect(editedValue).toBe("03.06.2023")
```
No `await` needed, because `toBe` does not return a promise.

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+    //   add new time boundaries
+    const newDateInput = page.locator('[data-testid="new-date-input-field"]')
+    await newDateInput.fill("01-05-2023")
+    await newDateInput.fill("02-06-2023")
```
Did you fill by using dashes (01-05-2023) on purpose to check that this works? My first intuition would be to start with the happy case with the correct German dot format (01.05.2023). I'd put input validation/rejection/acceptance in a separate test.

hamo225 replied:
good point, lets revisit when we add validation

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+    await expect(dateInputs).toHaveCount(3)
+    //edit time boundaries
+
+    await expect(dateInputs).toHaveCount(3)
```
Duplication of line 54?

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+
+    await page.waitForResponse(
+      (response) =>
+        response.url().includes("timeBoundaries") && response.status() === 200,
```
As you already have the response at hand, you could check for string equality with the expected response body: `[{"date":"2017-03-16","eventRefEid":"meta-1_lebzykl-1_ereignis-2"},{"date":"2023-05-01","eventRefEid":"meta-1_lebzykl-1_ereignis-3"},{"date":"2023-06-02","eventRefEid":"meta-1_lebzykl-1_ereignis-4"}]`

hamo225 replied:
Removed this. The playwright e2e tests should have this focus of what we expect a user to experience. Meaning we do not really need to test the payloads. I have instead added checks that the inputs display the expected values.

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+      .nth(1)
+    await dateInputToEdit.fill("03.06.2023")
+    await saveButton.click()
+    await page.reload()
```
Why here no `waitForResponse` like you did before? Was that not the reason to prevent a race condition -> flaky tests? Btw could check for response body `[{"date":"2017-03-16","eventRefEid":"meta-1_lebzykl-1_ereignis-2"},{"date":"2023-06-03","eventRefEid":"meta-1_lebzykl-1_ereignis-3"},{"date":"2023-06-02","eventRefEid":"meta-1_lebzykl-1_ereignis-4"}]`

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+      const deleteButton = page.locator(`[data-testid="delete-button-${i}"]`)
+      await deleteButton.click()
+      await saveButton.click()
+      await page.reload()
```
Same `waitForResponse` as above.

1st expectation: `[{"date":"2017-03-16","eventRefEid":"meta-1_lebzykl-1_ereignis-2"},{"date":"2023-06-03","eventRefEid":"meta-1_lebzykl-1_ereignis-3"}]`

2nd expectation `[{"date":"2017-03-16","eventRefEid":"meta-1_lebzykl-1_ereignis-2"}]`

Of course then this nice loop wouldn't work anymore :D

hamo225 replied:
we can discuss

reckseba commented on frontend/e2e/navigate-to-temporal-data-page-and-test-content.spec.ts:
```diff
@@ -1,34 +1,96 @@
+    const isDisabled = await deleteButton.isDisabled()
+    expect(isDisabled).toBe(true)
+
+    await page.request.put(
```
What about this PUT request? You are not expecting anything from it. And it does not change anything on the state.

hamo225 replied:
The PUT is not part of the user interaction. Its for the test, to set up and reset the environment. I have moved this into a function that will be called at the start and end of the test. This whole setup for the tests will change soon i hope.

hamo225 committed:
Updates e2e tests

- removes await with toBe assertion as does not return a promise

- removed the waitForResponse as not needed

- created a setup function that to be used at the start and finish of the test

RISDEV-3848

hamo225 committed:
Adds url for test page

RISDEV-3848
